lib/datadog/di/probe_notification_builder.rb

```diff
@@ -200,6 +202,18 @@ def get_local_variables(trace_point)
           map[name] = value
         end
       end
+
+      def active_trace
+        if defined?(Datadog::Tracing)
+          Datadog::Tracing.active_trace
+        end
+      end
+
+      def active_span
+        if defined?(Datadog::Tracing)
+          Datadog::Tracing.active_span
+        end
+      end
```

Comment:
Not sure what's the overhead trade-off (maybe it's very minor), but the `defined?(Datadog::Tracing)` check could be made inside the component (or init) and cached.

This is because anyway any change to load/enable a product means that the components will get rebuilt, so you'll always be able to check once and rely on that check from them on.

Reply:
I wrote a benchmark comparing a `defined?` call with instance variable retrieval (`@var`) and the `defined?` call is 66% of IPS of the instance variable retrieval, which I think is not a significant performance penalty.

I can make notification builder take a flag indicating whether to use tracing or not, which will be passed from a higher level class after it checks whether tracing is defined, for the purpose of explicitly controlling tracer usage within the notification builder (with the downside of needing to re-instantiate the notification builder when tracing is loaded, which practically I don't think will ever happen). But making this change for the purpose of optimization I think is not worthwhile.

What do you think? @marcotc @ivoanjo

Reply:
Thanks for checking... I think your findings support the `defined` version in terms of overhead.

This did get me thinking: I think in almost all cases the tracing code will be loaded, so if you are targeting customers that disable tracing, you may want to detect if tracing is enabled or not, rather than loaded or not.

Reply:
From what I understand, in practice every DI customer will be using tracing. I do not intend to specifically invest effort into supporting Di-without-tracing scenario, the only important part is I need to not load tracing earlier than it otherwise would be loaded due to integrations.

Reply:
In that case... I'm actually not sure this change is needed at all. Specifically, I understand the changes regarding the early boot of the library, but the `ProbeNotificationBuilder` only gets called via the components, and well, if the components are running, then tracing has already been required.

So I suspect this part of the change is a bit of a no-op?

Reply:
I don't want to assume that tracing exists if I (in DI) never required it (assuming that the library overall loaded tracing after it loaded DI).

Ideally the DI unit tests won't be loading tracing also (currently they do by virtue of loading the global spec helper).

Reply:
Yeah, I think that's OK, just wanted to clarify this bit since it seemed relevant context to how overall the components and whatnot works today :)